### ISSUE TITLE
Support retain_graph=True in PyTorch autograd integration

### ISSUE THREAD

**Issue by jhelferty-nv** [shader-slang/slangpy#914]
## Summary

SlangPy's `TorchAutoGradHook.backward()` does not support `retain_graph=True`. If a user calls `backward(retain_graph=True)` on a graph containing SlangPy operations and then calls `backward()` again on the same graph, the behavior is undefined — the native `autograd_backward` mutates diff pairs in-place (restoring primals, assigning gradients), so a second backward call would operate on already-modified state.

## Reproducer

```slang
// example.slang
import slangpy;

[Differentiable]
float square(float x) { return x * x; }
```

```python
import slangpy as spy
import torch

device = spy.create_torch_device(type=spy.DeviceType.cuda)
module = spy.Module.load_from_file(device, "example.slang")

x = torch.tensor([1.0, 2.0, 3.0], device="cuda", requires_grad=True)

y = module.square(x=x)
loss = y.sum()

# First backward with retain_graph=True
loss.backward(retain_graph=True)
print(x.grad)  # Expected: [2.0, 4.0, 6.0]

x.grad.zero_()

# Second backward on the same graph — currently undefined behavior
loss.backward()
print(x.grad)  # Expected: [2.0, 4.0, 6.0] (same gradients)
```

## Use cases for `retain_graph=True`

- GAN training (discriminator and generator losses sharing part of the graph)
- Higher-order gradients (gradients of gradients)
- Multiple loss functions calling `backward()` separately on shared intermediate computations

## Possible approaches

- Snapshot/restore the diff pair state so a second backward call can replay it
- Detect `retain_graph=True` proactively (PyTorch does not currently expose this flag to `Function.backward()`)

## Notes

- Neither SlangPy nor slang-torch currently use `retain_graph=True` anywhere
- Changes would likely be needed on both the Python side (`autogradhook.py`) and the C++ side (`NativeCallData::autograd_backward`)

## Interaction with #896

#896 proposes fixing a VRAM memory leak by releasing `ctx` attributes (`function`, `forwards_cd`, `args`, `kwargs`, `pairs`) after `backward()` completes. That fix would cause a second `backward()` call to fail with an `AttributeError` on the now-`None` attributes. If #896 is fixed first, the error should be caught explicitly with a clear message explaining that `retain_graph=True` is not supported. Supporting `retain_graph=True` would require preserving (or restoring) those attributes rather than clearing them.
